_posts/2016-11-09-composer-mailchimp-subscriptions-v3-api.md

```diff
@@ -0,0 +1,163 @@
+
+We implemented the following function, which can be reviewed and modified for your specific purposes:
+
+```php
+<?php
+/**
+ * Add an email to a MailChimp list.
+ *
+ * This code is based on the 7.x-4.6 version of the Mailchimp module,
+ * specifically the mailchimp_subscribe_process() function. That version of
+ * the Mailchimp contrib module makes use of the ThinkShout PHP library for
+ * version 3.0 of the MailChimp API. See the following for more detail:
+ * https://www.drupal.org/project/mailchimp
+ * https://github.com/thinkshout/mailchimp-api-php.
+ *
+ * @see Mailchimp_Lists::subscribe()
+ *
+ * @param string $api_key
+ *   The MailChimp API key.
+ * @param string $list_id
+ *   The MailChimp list id that the user should be subscribed to.
+ * @param string $email
+ *   The email address for the user being subscribed to the mailing list.
+ */
+function mymodule_subscribe_user($api_key, $list_id, $email) {
+
+  try {
+    // Set the timeout to something that won't take down the Drupal site:
+    $timeout = 60;
+    // Get an instance of the MailchimpLists class.
+    $mailchimp = new \Mailchimp\MailchimpLists($api_key, 'apikey', $timeout);
+
+    // Use MEMBER_STATUS_PENDING to require double opt-in for the subscriber. Otherwise, use MEMBER_STATUS_SUBSCRIBED.
+    $parameters = array(
+      'status' => \Mailchimp\MailchimpLists::MEMBER_STATUS_PENDING,
+      'email_type' => 'html',
+    );
+
+    // Subscribe user to the list.
+    $result = $mailchimp->addOrUpdateMember($list_id, $email, $parameters);
+
+    if (isset($result->id)) {
+      watchdog('mymodule', '@email was subscribed to list @list.',
+        array('@email' => $email, '@list' => $list_id), WATCHDOG_NOTICE
+      );
+    }
+    else {
+      watchdog('mymodule', 'A problem occurred subscribing @email to list @list.', array(
+        '@email' => $email,
+        '@list' => $list_id,
+      ), WATCHDOG_WARNING);
+    }
+  }
+  catch (Exception $e) {
+    // The user was not subscribed so log to watchdog.
+    watchdog('mymodule', 'An error occurred subscribing @email to list @list. Status code @code. "%message"', array(
+      '@email' => $email,
+      '@list' => $list_id,
+      '%message' => $e->getMessage(),
+      '@code' => $e->getCode(),
+    ), WATCHDOG_ERROR);
+  }
+}
+```
+
+With that function defined, we can then subscribe an email address to a specific Mailchimp mailing list through the following function call in our custom module:
+
+```php
+mymodule_subscribe_user($api_key, $list_id, $email);
+```
+
+## Conclusion
+
+By taking advantage of the modern PHP ecosystem built on reusable Composer managed packages, we can easily build or adapt a custom module to subscribe users to mailing lists without the MailChimp contributed module.
+
+Lastly, a special thanks to Thinkshout for their hard work maintaining the MailChimp module and creating the library, on which this approach depends!
```

Comment:
Thinkshout = ThinkShout -> to keep consistent with their name. Maybe a link to their homepage is appropriate?